Fixes #18687

galibzon opened the pull request:
## What does this PR do?
Fixes #18687 

`MaterialComponentController::OnSystemTick()` was missing the
additional check that `m_notifiedMaterialAssets` must be empty in order to call
`SystemTickBus::BusDisconnect()`.

Also added `GetMaterialMapCopy()` function to `MaterialComponentRequestBus`
so Lua scripts can actually make a copy of the current MaterialMap.

## How was this PR tested?

Validated on Win11 with Lua script that follows the steps as reported by Reverb. The Lua Script is available as an attachment in GHI #18687 .

gadams3 reviewed (commented):
My suspicion is that the issues relate to the recently added queue that defers material initialization from the asset bus notification handlers to the system tick. I’m not exactly sure how that addresses the streaming image-related deadlock or why it wasn’t handled directly within that system. I would think that if the issue were related to images, any other system using images would also experience the deadlock under similar conditions.

That being said, are there any logs, warnings, or errors reported for the original issue mentioned on Discord? I recall an issue with assigning material properties to an instance on the same tick it was initialized, which was prevented when initialization was handled in the asset bus handler. If that was the case, the system tick function may need to return after processing the notifier queue or reloading queued materials.

gadams3 reviewed (changes requested):
The copy function looks fine but I explained the issues with the other changes in a separate comment.

galibzon commented:
> My suspicion is that the issues relate to the recently added queue that defers material initialization from the asset bus notification handlers to the system tick. I’m not exactly sure how that addresses the streaming image-related deadlock or why it wasn’t handled directly within that system. I would think that if the issue were related to images, any other system using images would also experience the deadlock under similar conditions.
> 
> That being said, are there any logs, warnings, or errors reported for the original issue mentioned on Discord? I recall an issue with assigning material properties to an instance on the same tick it was initialized, which was prevented when initialization was handled in the asset bus handler. If that was the case, the system tick function may need to return after processing the notifier queue or reloading queued materials.

Explanation of how the new code fixes deadlock:
```
** Main Thread                | ** Secondary Copy Queue Thread
AssetBus::lock(mutex)         |
AssetBus::OnAssetReady        |
StreamingImage::FindOrCreate  |
AsyncUploadQueue::queueWork   |
Wait For Work Complete        |
                              |
                              | workQueue signaled
                              | Pop Work
                              | StreaminImage::Destructor()
                              | AssetBus::Disconnect()
                              | AssetBus::lock(mutex) <- Deadlocked
```

There are no Warnings or Errors at all. It just doesn't send the update signal to the Mesh Component. One of the reasons is because I noticed the code in `MaterialComponentController` only calls `AZ::Data::AssetBus::MultiHandler::BusConnect(uniqueMaterial.GetId());` and it never calls `BusDisconnect`. So, when `SetMaterialMap` is called, all the MaterialAssets are already Ready (and the controller is already `BusConnected`. So when it calls BusConnect(assetId) it doesn't trigger reloading because it was already connected and the update signal is never sent.

galibzon commented:
@gadams3 , I was able to find the missing piece, and this new version now checks `m_notifiedMaterialAssets` must be empty before calling `SystemTickBus::BusDisconnect()`. So now the core logic is simpler and just like before.

Also, I left a couple of TODO notes that we still need to make sure to AssetBus::MultiHandler::BusDisconnect(assetId) when a material asset is not used anymore.

nick-l-o3de commented on Gems/AtomLyIntegration/CommonFeatures/Code/Source/Material/MaterialComponentController.cpp:
```diff
@@ -130,6 +131,7 @@ namespace AZ
 
         void MaterialComponentController::Deactivate()
         {
+            AZ::Data::AssetBus::MultiHandler::BusDisconnect();
             MaterialComponentRequestBus::Handler::BusDisconnect();
             MaterialConsumerNotificationBus::Handler::BusDisconnect();
 
```
What happened to the whitespace here?  Sometimes its just a github glitch of course.

galibzon replied:
Maybe a glitch? I don't see any extra whitespace.

nick-l-o3de commented on Gems/AtomLyIntegration/CommonFeatures/Code/Source/Material/MaterialComponentController.cpp:
```diff
@@ -568,6 +583,13 @@ namespace AZ
                 // If the asset ID is different than what's already assigned then replace it
                 if (materialIt->second.m_materialAsset.GetId() != materialAssetId)
                 {
+                    // TODO: Although this works, we need to make sure that the previous material AssetId
+                    // is not being used in other slots before Disconnecting from the AssetBus.
+                    //const auto& prevAssetId = materialIt->second.m_materialAsset.GetId();
+                    //if (prevAssetId.IsValid())
+                    //{
+                    //    AZ::Data::AssetBus::MultiHandler::BusDisconnect(prevAssetId);
+                    //}
```
Is there a ticket for this or is your intention to follow up immediatel?

galibzon replied:
I'll follow up immediately, but need to consult with @gadams3.

nick-l-o3de commented on Gems/AtomLyIntegration/CommonFeatures/Code/Source/Material/MaterialComponentController.cpp:
```diff
@@ -560,6 +568,13 @@ namespace AZ
                     materialIt->second.m_propertyOverrides.empty() &&
                     materialIt->second.m_matModUvOverrides.empty())
                 {
+                    // TODO: Although this works, we need to make sure that the previous material AssetId
+                    // is not being used in other slots before Disconnecting from the AssetBus.
+                    //const auto& prevAssetId = materialIt->second.m_materialAsset.GetId();
+                    //if (prevAssetId.IsValid())
+                    //{
+                    //    AZ::Data::AssetBus::MultiHandler::BusDisconnect(prevAssetId);
+                    //}
```
Todos, immediate follow up or ticket?

galibzon replied:
I'll follow up immediately, but need to consult with @gadams3.

gadams3 commented:
> > My suspicion is that the issues relate to the recently added queue that defers material initialization from the asset bus notification handlers to the system tick. I’m not exactly sure how that addresses the streaming image-related deadlock or why it wasn’t handled directly within that system. I would think that if the issue were related to images, any other system using images would also experience the deadlock under similar conditions.
> > That being said, are there any logs, warnings, or errors reported for the original issue mentioned on Discord? I recall an issue with assigning material properties to an instance on the same tick it was initialized, which was prevented when initialization was handled in the asset bus handler. If that was the case, the system tick function may need to return after processing the notifier queue or reloading queued materials.
> 
> Explanation of how the new code fixes deadlock:
> 
> ```
> ** Main Thread                | ** Secondary Copy Queue Thread
> AssetBus::lock(mutex)         |
> AssetBus::OnAssetReady        |
> StreamingImage::FindOrCreate  |
> AsyncUploadQueue::queueWork   |
> Wait For Work Complete        |
>                               |
>                               | workQueue signaled
>                               | Pop Work
>                               | StreaminImage::Destructor()
>                               | AssetBus::Disconnect()
>                               | AssetBus::lock(mutex) <- Deadlocked
> ```
> 
> There are no Warnings or Errors at all. It just doesn't send the update signal to the Mesh Component. One of the reasons is because I noticed the code in `MaterialComponentController` only calls `AZ::Data::AssetBus::MultiHandler::BusConnect(uniqueMaterial.GetId());` and it never calls `BusDisconnect`. So, when `SetMaterialMap` is called, all the MaterialAssets are already Ready (and the controller is already `BusConnected`. So when it calls BusConnect(assetId) it doesn't trigger reloading because it was already connected and the update signal is never sent.

Sorry if I wasn't clear, I was referring to logs, warnings, or errors related to the issue mentioned in discord about materials not being assigned or updated.

Regarding the deadlock, I was just trying to understand why or how this problem was specific to the material component instead of being resolved in a more generic way related to streaming images.

Regarding the comment about not disconnecting from the asset bus, maintaining a connection to the asset bus is required to receive asset reload notifications for hot reloading. It should be disconnecting from the bus for all IDs whenever it attempts to reload after any material asset assignment changes or on deactivation.

gadams3 commented:
> @gadams3 , I was able to find the missing piece, and this new version now checks `m_notifiedMaterialAssets` must be empty before calling `SystemTickBus::BusDisconnect()`. So now the core logic is simpler and just like before.
> 
> Also, I left a couple of TODO notes that we still need to make sure to AssetBus::MultiHandler::BusDisconnect(assetId) when a material asset is not used anymore.

That's good news. I can see the change. I think that other block still needs to be reverted back to that single function call. See my other comments about disconnecting from the asset bus. It does get disconnected on deactivation and whenever release materials is called, which is the first thing done before loading materials. Using the original logic, the asset bus handler should only ever be connected to materials directly referenced by the component, once release or load materials is called, usually on tick.

galibzon commented:
> > > My suspicion is that the issues relate to the recently added queue that defers material initialization from the asset bus notification handlers to the system tick. I’m not exactly sure how that addresses the streaming image-related deadlock or why it wasn’t handled directly within that system. I would think that if the issue were related to images, any other system using images would also experience the deadlock under similar conditions.
> > > That being said, are there any logs, warnings, or errors reported for the original issue mentioned on Discord? I recall an issue with assigning material properties to an instance on the same tick it was initialized, which was prevented when initialization was handled in the asset bus handler. If that was the case, the system tick function may need to return after processing the notifier queue or reloading queued materials.
> > 
> > 
> > Explanation of how the new code fixes deadlock:
> > ```
> > ** Main Thread                | ** Secondary Copy Queue Thread
> > AssetBus::lock(mutex)         |
> > AssetBus::OnAssetReady        |
> > StreamingImage::FindOrCreate  |
> > AsyncUploadQueue::queueWork   |
> > Wait For Work Complete        |
> >                               |
> >                               | workQueue signaled
> >                               | Pop Work
> >                               | StreaminImage::Destructor()
> >                               | AssetBus::Disconnect()
> >                               | AssetBus::lock(mutex) <- Deadlocked
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > There are no Warnings or Errors at all. It just doesn't send the update signal to the Mesh Component. One of the reasons is because I noticed the code in `MaterialComponentController` only calls `AZ::Data::AssetBus::MultiHandler::BusConnect(uniqueMaterial.GetId());` and it never calls `BusDisconnect`. So, when `SetMaterialMap` is called, all the MaterialAssets are already Ready (and the controller is already `BusConnected`. So when it calls BusConnect(assetId) it doesn't trigger reloading because it was already connected and the update signal is never sent.
> 
> Sorry if I wasn't clear, I was referring to logs, warnings, or errors related to the issue mentioned in discord about materials not being assigned or updated.
> 
> Regarding the deadlock, I was just trying to understand why or how this problem was specific to the material component instead of being resolved in a more generic way related to streaming images.
> 
> Regarding the comment about not disconnecting from the asset bus, maintaining a connection to the asset bus is required to receive asset reload notifications for hot reloading. It should be disconnecting from the bus for all IDs whenever it attempts to reload after any material asset assignment changes or on deactivation.

About the deadlock...

Yes, it can occur on any piece of code that instantiates StreamingImages under the scope of AssetBus::OnAssetReady. Luckily there are only a few components that required the `PostTick` fix.

Now, why it did not occurr before? When did it become a problem? It started with this improvement from Huawei (https://github.com/o3de/o3de/pull/18479), where there were dangling raw pointer to Image Resources, that now became Ref Counted, and which Triggered the possibility of StreamingImage instances to be released during AsyncUploadQueue Thread. Before there was no deadlock, but instead there were crashes as explained in #18479 ).

Fixing the deadlock inside StreamingImages could require major re-architecture. Maybe it can be fixed with a `StreamingImage Garbage Collector` Queue Service. Each time a StreamingImage needs to be released it should be added to the Queue, and all items collected in the Queue are released on the main thread by the Service.

gadams3 commented:
> Can you show me the line where `Data::AssetBus::MultiHandler::BusDisconnect(assetId)` is called (Besides the one I just added)? Doing a search in the code, I don't see it.

Disconnect is called in a couple of locations without passing it a specific ID, which should disconnect from all IDs. Then it explicitly reconnects to valid ID's before attempting to reload them. Step through load materials, which calls release materials, which disconnects completely from the asset bus, before load materials reconnects specific material asset IDs.

galibzon commented:
> > Can you show me the line where `Data::AssetBus::MultiHandler::BusDisconnect(assetId)` is called (Besides the one I just added)? Doing a search in the code, I don't see it.
> 
> Disconnect is called in a couple of locations without passing it a specific ID, which should disconnect from all IDs. Then it explicitly reconnects to valid ID's before attempting to reload them. Step through load materials, which calls release materials, which disconnects completely from the asset bus, before load materials reconnects specific material asset IDs.

Maybe you are referring to a very old version of MaterialComponentController? I have done extensive debugging the last few days and MaterialComponentController never calls BusDisconnect() (Which I just added in `Deactivate()`).

gadams3 commented on Gems/AtomLyIntegration/CommonFeatures/Code/Source/Material/MaterialComponentController.cpp:
```diff
@@ -130,6 +131,7 @@ namespace AZ
 
         void MaterialComponentController::Deactivate()
         {
+            AZ::Data::AssetBus::MultiHandler::BusDisconnect();
```
Deactivate calls ReleaseMaterials three lines below this.

        void MaterialComponentController::ReleaseMaterials()
        {
            SystemTickBus::Handler::BusDisconnect();
            Data::AssetBus::MultiHandler::BusDisconnect();

galibzon replied:
Yeah, I got confused between using two computers (at different commits) and my redundant call using`AZ::`, and my text search was using `AZ::*`, etc. All good now and I removed the redundant BusDisconnect.
